pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,387 @@
+	configv1.OperatorDegraded:    configv1.ConditionFalse,
+}
+
+func (f *frontend) postAdminOpenShiftClusterEtcdCertificateRenew(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	err := f._postAdminOpenShiftClusterEtcdCertificateRenew(ctx, r, log)
+
+	adminReply(log, w, nil, nil, err)
+}
+
+// validate cluster is <4.9 and etcd is in expected state
+func (e *etcdrenew) validateAndBackup(ctx context.Context, takeBackup bool) error {
+	s := []steps.Step{
+		steps.Action(e.validateEtcdOperatorControllersState),
+		steps.Action(e.validateEtcdOperatorState),
+		steps.Action(e.validateEtcdCertsExistsAndExpiry),
+	}
+	if takeBackup {
+		s = append(s,
+			steps.Action(e.backupEtcdSecrets),
+			steps.Action(e.deleteEtcdSecrets),
+		)
+	}
+
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *etcdrenew) isRenewed(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Condition(e.isRevised, 30*time.Minute, true),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *frontend) _postAdminOpenShiftClusterEtcdCertificateRenew(ctx context.Context, r *http.Request, log *logrus.Entry) error {
+	resourceName := chi.URLParam(r, "resourceName")
+	resourceType := chi.URLParam(r, "resourceType")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", resourceType, resourceName, resourceGroupName)
+	case err != nil:
+		return err
+	}
+
+	k, err := f.kubeActionsFactory(log, f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+	e := &etcdrenew{
+		log:           log,
+		k:             k,
+		secretNames:   nil,
+		mode:          "renew",
+		backupSecrets: make(map[string][]byte),
+		lastRevision:  0,
+	}
+
+	if err = e.validateClusterVersion(ctx); err != nil {
+		return err
+	}
+
+	// Fetch secretNames using nodeNames
+	masterNodeNames, err := fetchNodeNames(ctx, k, log)
+	if err != nil {
+		return err
+	}
+	if len(masterNodeNames) != 3 {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "The cluster doesn't have 3 master nodes")
+	}
+
+	for _, nodeName := range masterNodeNames {
+		for _, prefix := range []string{"etcd-peer-", "etcd-serving-", "etcd-serving-metrics-"} {
+			e.secretNames = append(e.secretNames, prefix+nodeName)
+		}
+	}
+
+	// validate, backup and delete secrets
+	if err = e.validateAndBackup(ctx, true); err != nil {
+		return err
+	}
+
+	if err = e.isRenewed(ctx); err != nil {
+		e.mode = "recovery"
```

Comment:
Having multiple "modes" like this generally leads to over-complexity, it's similar to old-timey `goto` calls - they're hard to follow, review for bugs, and maintain later, because for the rest of the review, I now have to think "what state am I in right now?" and "what states COULD I be in right now?". Generally, it's safer and easier to keep things stateless.

Can we restructure the code to not require modes, so it's more declarative right in this `if` what needs to happen for each case? For example, something pseudo like:

```
if err != nil {
    e.log.Println("Attempting Recovery of etcd cluster")
    // all the functions required to recover, maybe wrapped in a giant e.RecoverEtcdCluster() function or similar...
} else {
    e.log.Println("Attempting to renew etcd certificates")
    // all the functions required to renew certs, maybe wrapped in a giant e.RenewEtcdCertificates() function or similar...
}
e.log.Println("Done")